README.md

```diff
@@ -86,6 +86,71 @@ $ java -jar edge-sip2-fat.jar -conf '{"port":1234,"okapiUrl":"https://folio-snap
 
+## Metrics
+
+This module makes use of [Micrometer](https://micrometer.io) to collect SIP2, Vert.x and JVM metrics. The metrics need to be collected by a monitoring system backed. This is where Micrometer provides flexibility, by allowing the module to code to the Micrometer interface, which is vendor neutral. Once determined, the vendor specific backend binding is provided runtime and can be easily replaced.
+
+### Enabling Vert.x metrics
+
+By default, metrics are disabled. To enable Vert.x metrics pass the following Java argument:
+
+```
+-Dvertx.metrics.options.enabled=true
+```
+
+With metrics enabled, configuration must be supplied to the verticle. This can be done as follows:
+
+```
+ -options '{"metricsOptions":{"labels":["LOCAL","REMOTE","HTTP_PATH","HTTP_METHOD","HTTP_CODE","CLASS_NAME"],"enabled":true,"prometheusOptions":{"enabled":true,"startEmbeddedServer":true,"embeddedServerOptions":{"port":8081}}}}'
+```
+
+The `metricsOptions` here indicate that the verticle should collect metrics with the supplied list of `labels`. Some of these labels, like `REMOTE`, may lead to high cardinality metrics. The default `labels` list is "HTTP\_METHOD", "HTTP\_CODE", "POOL\_TYPE" and "EB\_SIDE".
+
+Also specified here are `prometheusOptions`. In this case, the backend will be [Prometheus](https://prometheus.io). Prometheus "scrapes" metrics via HTTP at a specified interval. The options specified here allow Vert.x to create an HTTP server to handle metrics scraping. Prometheus setup is outside the scope of this document. Other bindings could be used, like [InfluxDB](https://www.influxdata.com/products/influxdb-overview/).
+
+### Available metrics
+
+For a list of Vert.x metrics (HTTP Client and Net Server are the primary sources for metrics in this module) see: [Vert.x core tools metrics](https://vertx.io/docs/vertx-micrometer-metrics/java/#_vert_x_core_tools_metrics)
+
+The following metrics are supplied by this module:
+
+|Metric name|Labels|Type|Description|
+|-----------|------|----|-----------|
+|`org_folio_edge_sip2_command_timer`|`command`|Timer|SIP2 command execution time|
+|`org_folio_edge_sip2_invalidMessage_errors`|`port`|Counter|A count of invalid message errors|
+|`org_folio_edge_sip2_request_errors`|`port`|Counter|A count of request errors|
+|`org_folio_edge_sip2_response_errors`|`port`|Counter|A count of response errors|
+|`org_folio_edge_sip2_scResend_errors`|`port`|Counter|A count of SC resend errors, which occurs when the module fails to send the SC a resend message when the prior received message was not understood|
+|`org_folio_edge_sip2_socket_errors`|`port`|Counter|A count of socket errors|
+
```

Comment:
Do you want to enumerate a few (or all) counters that we talked about, such as rate and timer counters for each transaction, such as PatronInformation?

Reply:
I'm not sure what you mean here. The first line in the table is the command timer and that will have labels/tags for the command that it is measuring. There isn't explicit PatronInformation timer. All commands are `org_folio_edge_sip2_command_timer`.

Reply:
If there isn't explicit PatronInformation timer then the org_folio_edge_sip2_command_timer will tell the average execution time of all sip commands, but cannot tell us an individual command's time? When you showed me on Grafana I thought I was seeing a specific counter for PatroInformation. I may have misunderstood.

Reply:
Micrometer has the concept of "tags" which can be used to drill down for more details. In grafana the metric we were looking at was "org_folio_edge_sip2_command_timer_seconds_sum", but in the graph, thanks to the "command" tag, we were also looking at data aggregated for each "command", which was displayed as: `{command="PATRON_INFORMATION",instance="172.17.0.1:8081",job="vertx-8081"}`, sort of like select with a group by clause in SQL.

So, if any of these tags are different, than there would be another line on the graph. I believe the instance/job here is my Prometheus server, so that shouldn't change, but "command" will change for each SIP2 command received. It is similar for Vert.x's HTTP timers. We'll have to figure out what to do with redundancy... I think having command rates per edge-sip2 instance is fine, but we'll see.

Setting the command tag is why I used a `Timer.Sample` instead of a `Timer` in micrometer. The timer needs to know the tags upfront, as far as I could tell, and we wouldn't parse the command until after the timer was started. The sample allows for setting the tag at stop when we know the command.

I think we'll discover other cool things as we go and probably find out things that could be implemented better.

Does this make sense?

Reply:
sounds good